src/typesystem.rs

```diff
@@ -0,0 +1,338 @@
+
+    //TODO : Add location information
+}
+
+impl DataType {
+
+    pub fn get_name(&self) -> &str {
+        self.name.as_str()
+    }
+
+    pub fn get_type_information(&self) -> &DataTypeInformation {
+        &self.information
+    }
+
+    pub fn clone_type_information(&self) -> DataTypeInformation {
+        self.information.clone()
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum DataTypeInformation {
+    Struct {
+        name: String,
+        member_names: Vec<String>,
+    },
+    Array {
+        name : String,
+        inner_type_name: String,
+        dimensions : Vec<Dimension>, 
+    },
+    Integer {
+        signed: bool,
+        size: u32,
+    },
+    Float {
+        size: u32,
+    },
+    String {
+        size: u32,
+    }, 
+    Alias {
+        name: String,
+        referenced_type: String,
+    },
+    Void,
+}
+
+impl DataTypeInformation {
+
+    pub fn get_name(&self) -> &str {
+        match self {
+            DataTypeInformation::Struct { name, .. } => name,
+            DataTypeInformation::Array { name, .. } => name,
+            DataTypeInformation::Integer { size, ..} => get_integer_type_name(size), 
+            DataTypeInformation::Float { size, .. } => get_float_type_name(size), 
+            DataTypeInformation::String { ..} => "String", 
+            DataTypeInformation::Alias { name, .. } => name,
+            DataTypeInformation::Void  => "Void",
+        }
+    }
+
+    pub fn is_int(&self) -> bool {
+        if let DataTypeInformation::Integer { .. } = self {
+            true
+        } else {
+            false
+        }
+    }
+
+    pub fn is_float(&self) -> bool {
+        if let DataTypeInformation::Float { .. } = self {
+            true
+        } else {
+            false
+        }
+    }
+
+    pub fn is_numerical(&self) -> bool {
+        match self {
+            DataTypeInformation::Integer { .. } | DataTypeInformation::Float { .. } => true,
+            _ => false,
+        }
+    }
+
+    pub fn get_size(&self) -> u32 {
+        match self {
+            DataTypeInformation::Integer { size, .. } => *size,
+            DataTypeInformation::Float { size, .. } => *size,
+            DataTypeInformation::String { size, .. } => *size,
+            DataTypeInformation::Struct { .. } => 0, //TODO : Should we fill in the struct members here for size calculation or save the struct size.
+            DataTypeInformation::Array { .. } => unimplemented!(), //Propably length * inner type size
+            DataTypeInformation::Alias { .. } => unimplemented!(),
+            DataTypeInformation::Void => 0,
+        }
+    }
+}
+
+fn get_integer_type_name(size : &u32) -> &str{
+    match size {
+        1 => "BOOL",
+        8 => "BYTE",
+        16 => "INT",
+        32 => "DINT",
+        64 => "LINT",
+        _ => "Interger"
+    }
+}
+
+fn get_float_type_name(size : &u32) -> &str{
+    match size {
+        32 => "REAL",
+        64 => "LREAL",
+        _ => "Float"
```

Comment:
we gonna ask ourselfs in the future: "where the fuck does this 'Float' type come from!?"
shouldnt we fail here?

Reply:
Done